src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,204 @@
+        public void CopyChangedValues(ICollection<TStructuralType> original)
+        {
+            //Here we need to Find the key of the Type, then only we will be able to find from the collection that which item in the collection
+            //corresponds to the item in delta list(Edmchangedobjectcoll).
+            Type type = typeof(TStructuralType);
+            IEnumerable<IEdmStructuralProperty> keys = EntityType.Key();
+
+            foreach (dynamic changedObj in _items)
+            {
+                //Get filtered item based on keys
+                TStructuralType originalObj = GetFilteredItem(type, keys, original as IEnumerable<TStructuralType>, changedObj);
+                                                
+                IEdmDeltaDeletedEntityObject deletedObj = changedObj as IEdmDeltaDeletedEntityObject;
+
+                if (deletedObj != null)
+                {
+                    if (originalObj != null)
+                    {
+                        //This case handle deletions
+                        original.Remove(originalObj);
+                    }
+                }
+                else
+                {
+                    if (originalObj == null)
+                    {
+                        //This case handle additions
+                        originalObj = Activator.CreateInstance(changedObj.ExpectedClrType);
+                        original.Add(originalObj);
+                    }
+
+                    //Patch for addition/update. This will call Delta<T> for each item in the collection
+                    changedObj.Patch(originalObj);
+                }               
+            }
+        }
+
+        private static TStructuralType GetFilteredItem(Type type, IEnumerable<IEdmStructuralProperty> keys, IEnumerable<TStructuralType> originalList, IEdmChangedObject changedObject)
+        {   
+            //This logic is for filtering the object based on the set of keys,
+            //There will only be very few key elements usually, mostly 1, so performance wont be impacted.
+            foreach(IEdmStructuralProperty key in keys)
+            {
+                object obj;
+                if (changedObject.TryGetPropertyValue(key.Name, out obj))
+                {
+                    originalList = originalList.Where(x => type.GetProperty(key.Name).GetValue(x).ToString() == obj.ToString());
+                }
+            }
+
+            return originalList.SingleOrDefault();
+        }
+
+        /// <summary>
+        /// Patch for EdmChangedobjectCollection, a collection for Delta<typeparamref name="TStructuralType"/>
+        /// </summary>
+        /// <param name="original">Original collection of the Type which needs to be updated</param>        
+        public void Patch(ICollection<TStructuralType> original)
+        {
+            CopyChangedValues(original);
+        }
+
+        internal ICollection<TStructuralType> GetInstance()
```

Comment:
>GetInstance [](start = 46, length = 11)

What is this method for? It doesn't seem to be used anywhere.

Reply:
We should remove this method as it's internal and not being used and makes certain assumptions (like the collection is a collection of DeltaOfTStructuralType that implements GetInstance).


---
In reply to: [551641080](https://github.com/OData/WebApi/pull/2275#discussion_r551641080) [](ancestors = 551641080)